### PR TITLE
(Finished) Ruby - Advanced Building Blocks - Enumerable Methods 

### PR DESCRIPTION
**Pull request by Dandush03**
(Finished) Ruby - Advanced Building Blocks - Enumerable Methods

### CODE REVIEW

**Comment by Anna-Myzukina** [comment]
### Your project looks :+1:  good!
### Great job thus far. 

Kindly recommended checking all your methods. It seems that all methods are working not correctly now and give errors.
* For example





Take a look at Ruby docs about Enumerable class https://ruby-doc.org/core-2.6.5/Enumerable.html



screen

### Make changes and submit another code review request.

*If you have any questions, let me know on slack (@Anna-Muzykina) and I will be glad to clarify with you.*

**Comment by Dandush03** [comment]
Hey, Just Fixed the issue, I think. I will appreciate it if the problem persist if you can let me know how you doing to debug it since I really just guessing what's the problem. I will appreciate if you point me to the right direction

**Review by fatymahmed — COMMENTED** [review]
### Pull Request under the review by Fatima.

**Review by fatymahmed — CHANGES_REQUESTED** [review]
### Hi @Dandush03 
Great job on what you have done so far👏👏
- Most of your tests are failing.
Screenshot
As you can see from the image above, the common error is `no block given (yield)`
- If you have yield within your method definition, that means you have to obligatorily pass a block when you use it, that's why it's failing for the tests where no block is passed.
- For example for the `each_with_index` method, without passing a block, it returns an Enumerator.
`[2,3,4,5].each_with_index`
returns 
`#<Enumerator: [2, 3, 4, 5]:each_with_index>`
For your `my_each_with_index` method, when you run it without a block (just like the one above), it produces an error `no block given (yield) `, that is because it expects a block.
- Go through all your methods and make sure they cover both cases where a block is passed as an argument and the other case is when there's no block.
- Check [this](https://ruby-doc.org/core-2.6.5/Enumerable.html) to know the different arguments that can be passed to these methods so you can have all cases.
I hope it's clear now.

[OPTIONAL]
- For this project and all the next ones, you are supposed to use `Git Flow` as [mentioned on Pathwright](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/git_flow.png), check [this](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59882084/) for information on Git flow.

### Kindly make the suggested changes and request for another code review.
### Status: Changes Required.
_Feel free to reach out to me on slack (Fatima) If you have any questions, I'd be happy to help_

**Comment by Dandush03** [comment]
> ### Hi @Dandush03
> Great job on what you have done so far
> 
> * Most of your tests are failing.
>   Screenshot
>   As you can see from the image above, the common error is `no block given (yield)`
> * If you have yield within your method definition, that means you have to obligatorily pass a block when you use it, that's why it's failing for the tests where no block is passed.
> * For example for the `each_with_index` method, without passing a block, it returns an Enumerator.
>   `[2,3,4,5].each_with_index`
>   returns
>   `#<Enumerator: [2, 3, 4, 5]:each_with_index>`
>   For your `my_each_with_index` method, when you run it without a block (just like the one above), it produces an error `no block given (yield) `, that is because it expects a block.
> * Go through all your methods and make sure they cover both cases where a block is passed as an argument and the other case is when there's no block.
> * Check [this](https://ruby-doc.org/core-2.6.5/Enumerable.html) to know the different arguments that can be passed to these methods so you can have all cases.
>   I hope it's clear now.
> 
> [OPTIONAL]
> 
> * For this project and all the next ones, you are supposed to use `Git Flow` as [mentioned on Pathwright](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/git_flow.png), check [this](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59882084/) for information on Git flow.
> 
> ### Kindly make the suggested changes and request for another code review.
> ### Status: Changes Required.
> _Feel free to reach out to me on slack (Fatima) If you have any questions, I'd be happy to help_

Thanks! I think I got it know... The only issue left unresolved is the 
### Metrics/ModuleLength: Module has too many lines. [110/100]

**Comment by Dandush03** [comment]
> > ### Hi @Dandush03
> > Great job on what you have done so far
> > 
> > * Most of your tests are failing.
> >   Screenshot
> >   As you can see from the image above, the common error is `no block given (yield)`
> > * If you have yield within your method definition, that means you have to obligatorily pass a block when you use it, that's why it's failing for the tests where no block is passed.
> > * For example for the `each_with_index` method, without passing a block, it returns an Enumerator.
> >   `[2,3,4,5].each_with_index`
> >   returns
> >   `#<Enumerator: [2, 3, 4, 5]:each_with_index>`
> >   For your `my_each_with_index` method, when you run it without a block (just like the one above), it produces an error `no block given (yield) `, that is because it expects a block.
> > * Go through all your methods and make sure they cover both cases where a block is passed as an argument and the other case is when there's no block.
> > * Check [this](https://ruby-doc.org/core-2.6.5/Enumerable.html) to know the different arguments that can be passed to these methods so you can have all cases.
> >   I hope it's clear now.
> > 
> > [OPTIONAL]
> > 
> > * For this project and all the next ones, you are supposed to use `Git Flow` as [mentioned on Pathwright](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/git_flow.png), check [this](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59882084/) for information on Git flow.
> > 
> > ### Kindly make the suggested changes and request for another code review.
> > ### Status: Changes Required.
> > _Feel free to reach out to me on slack (Fatima) If you have any questions, I'd be happy to help_
> 
> Thanks! I think I got it know... The only issue left unresolved is the
> 
> ### Metrics/ModuleLength: Module has too many lines. [110/100]

Never mind finding the solution. just splited module jeje

File: main.rb [hunk 1]
```diff
@@ -1,26 +1,266 @@
 module Enumerable
   def my_each
```
Comment (Forison) [inline main.rb:2]: * [x] should have working `my_each` identical and returns the same thing as ruby's [`each`](https://ruby-doc.org/core-2.6.4/Array.html#method-i-each).([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_each.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
``` p [1,2,3,4,5].my_each{ |a| a} != p [1,2,3,4,5].each{ |a| a}```

File: main.rb [hunk 2]
```diff
@@ -1,26 +1,266 @@
...
       while min < max
         yield(self[min])
         min += 1
       end
+    else
+      to_enum(:my_each)
+    end
+  end
+
+  def my_each_with_index
```
Comment (Forison) [inline main.rb:25]: * [ ]should have working `my_each_with_index` identical and returns the same thing as ruby's [`each_with_index`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-each_with_index). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_each_with_index.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
let us improve this method to work like it original 
```p [1,2,3,4,5].my_each_with_index{|a| a} != [1,2,3,4,5].each_with_index{|a| a}```

File: main.rb [hunk 3]
```diff
@@ -1,26 +1,266 @@
...
+      my_each do |item|
+        items << item if yield(item)
+      end
+      items
+    else
+      to_enum(:my_select)
+    end
+  end
+
+  def my_all?
```
Comment (Forison) [inline main.rb:61]: * [ ] should have working `my_all?` identical and returns the same thing as ruby's [`all?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-all-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_all.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

```  p %w[ant bear cat].all?(/t/)  != p %w[ant bear cat].my_all?(/t/) ```
```  p [1, 2i, 3.14].all?(Numeric) != p [1, 2i, 3.14].my_all?(Numeric)```
```  p [nil, true, 99].all? != p [nil, true, 99].my_all?```

File: main.rb [hunk 4]
```diff
@@ -1,26 +1,266 @@
...
+    bool_item = true
+    if block_given?
+      my_each do |item|
+        bool_item = false unless yield(item)
+      end
     end
+    bool_item
   end
+
+  def my_any?
```
Comment (Forison) [inline main.rb:71]: * [x] should have working `my_any?` identical and returns the same thing as ruby's [`any?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-any-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_any.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2)) 
``` p %w[ant bear cat].my_any?(/d/) !=  p %w[ant bear cat].any?(/d/)```
 ```p [nil, true, 99].my_any?(Integer) != p [nil, true, 99].any?(Integer)```
 ``` p [nil, true, 99].my_any? != p [nil, true, 99].any?``` 
 ``` p [].my_any? != p [].any? ```

File: main.rb [hunk 5]
```diff
@@ -1,26 +1,266 @@
...
+      my_each do |item|
+        bool_item = true if yield(item)
+      end
+    end
+    bool_item
+  end
+end
+
+module Enumerable
+  def my_none?
```
Comment (Forison) [inline main.rb:84]: * [ ] should have working `my_none?` identical and returns the same thing as ruby's [`none?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-none-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_none.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
compare the results of these examples with their corresponding original method. like the above example
```
p %w{ant bear cat}.my_none?(/d/)                     
p [1, 3.14, 42].my_none?(Float)  
p [].my_none?                        
p [nil].my_none?                       
p [nil, false].my_none? 
```

File: main.rb [hunk 6]
```diff
@@ -1,26 +1,266 @@
...
+      my_each do |item|
+        is_proc ? new_items.push(is_proc.call(item)) : new_items << yield(item)
+      end
+      new_items
+    else
+      to_enum(:my_map)
+    end
+  end
+
+  def my_inject(temp_option = nil)
```
Comment (Forison) [inline main.rb:120]: * [ ] should have working `my_inject` identical and returns the same thing as ruby's [`inject`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_inject.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
```p (5..10).my_inject { |sum, n| sum + n } != p (5..10).inject { |sum, n| sum + n }```
```p (5..10).my_inject(1) { |product, n| product * n } != p (5..10).inject(1) { |product, n| product * n }```

File: main.rb [hunk 7]
```diff
@@ -1,26 +1,266 @@
...
+    new_items = temp_option || shift
+    my_each do |item|
+      new_items = yield(new_items, item)
+    end
+    new_items
+  end
+end
+
+def multiply_els(array_num)
+  array_num.my_inject { |results, number| results * number }
```
Comment (Forison) [inline main.rb:130]: The change above will also affect this method as it depends on ```my_inject```

**Review by Forison — CHANGES_REQUESTED** [review]
### Status Require: changes🛑
# 🙋‍♂️@Dandush,
Good attempts on this one🙂. 
Here are the issues to address to get the project meet the requirement🙂.
* [ ] Files not related to the project should not be committed and added to .gitignore ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/general/gitignore.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/66243643/)) ```main.rb.swo and main.rbs.wp``` must not be commited.
## [_Optional_]
* [ ] The required Git Flow: production `master` branch + main `development` branch + Pull Requests from feature branches to the `develop` ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/git_flow.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/57421589/))
## Check below 👇for other mandatory requirements.
Kindly address all these changes and submit a new code review
### Happy coding💻
#### Addo Forison

**Pull request closed without merging**